net/interfaces/interfaces.go

```diff
@@ -117,17 +121,39 @@ func LocalAddresses() (regular, loopback []netaddr.IP, err error) {
 				if tsaddr.IsTailscaleIP(ip) {
 					continue
 				}
-				if ip.IsLinkLocalUnicast() {
-					continue
-				}
 				if ip.IsLoopback() || ifcIsLoopback {
 					loopback = append(loopback, ip)
+				} else if ip.Is4() && ip.IsLinkLocalUnicast() {
+					// In AWS Lambda, NAT for 169.254.x.x is the
+					// only connectivity.
+					linklocal4 = append(linklocal4, ip)
+
+					// We know of no cases where the IPv6 fe80:: addresses
+					// are used to provide WAN connectivity. It is also very
+					// common for users to have no IPv6 WAN connectivity,
+					// but their OS supports IPv6 so they have an fe80::
+					// address. We don't report IPv6 LL to Control.
+				} else if ip.Is6() && tsaddr.IsULA(ip) {
+					// Google Cloud Run uses NAT with IPv6 Unique Link
+					// Local addresses to provide IPv6 connectivity.
+					linklocal6 = append(linklocal6, ip)
 				} else {
-					regular = append(regular, ip)
+					if ip.Is4() {
+						regular4 = append(regular4, ip)
+					} else {
+						regular6 = append(regular6, ip)
+					}
 				}
 			}
 		}
 	}
+	if len(regular4) == 0 {
```

Comment:
this mixing regular & linklocal across IPv4 vs IPv6 still weirds me out. Does it suffice for the two environments you care about to only blend in linklocal if both regular4 and regular6 are both empty?

Reply:
Yes. Done.